### ISSUE TITLE
Product Query: [Filters] When inside a Product by Category template using a Filter by Attribute, non-pertaining attributes can be selected

### ISSUE THREAD

**Issue by sunyatasattva** [woothemes/woocommerce#42542]
> **Warning**
> In order to reproduce this bug, woocommerce/woocommerce-blocks#7382 is suggested

## Describe the bug

If a “Filter by Attribute” block is added within a “Products by Category” template, it shows all attributes available even if not pertaining to that category.

In combination with the normal classic block, it produces a mismatch between the data shown in the filter block and the classic block (similar, but different from https://github.com/woocommerce/woocommerce-blocks/issues/7607).

When combined the “Product Query” block with the “inherit query from template option”, it allows products from different categories to be displayed inside another category (e.g. a yellow cap inside the t-shirts category, as shown in the screenshot).

Also be aware of https://github.com/woocommerce/woocommerce-blocks/issues/7604 when reproducing this issue.

## To reproduce

Steps to reproduce the behavior with WooCommerce default sample data:

1. Add a “Filter by Attribute” block to the “Products by Category” template.
2. Go to a specific category, such as “T-shirts”.
3. Notice that the filter block shows attributes not available in the “T-shirts” category, such as “Yellow (1)”.

Optionally:

1. Make sure woocommerce/woocommerce-blocks#7382 is merged.
2. Add a “Product Query” block with “inherit query from template” option.
3. Repeat the steps above and select “Yellow”.
4. Notice you'll have a cap inside the t-shirts category.

## Screenshots

![Screenshot 2022-11-08 at 02 26 51](https://user-images.githubusercontent.com/1847066/200451225-21b53788-3686-4424-a60c-ff0f0bcbc1df.png)

**Comment by sunyatasattva** [comment]
Related to https://github.com/woocommerce/woocommerce-blocks/issues/7607

**Comment by OC2PS** [comment]
There's a "needs feedback" tag on this ticket, but I don't see any unanswered queries. Could Lucio @sunyatasattva or someone please clarify?

**Comment by sunyatasattva** [comment]
Hello @OC2PS yes, sorry: we are investigating this internally and there has been a lot of discussion and research on Database queries and how to solve this issue. That's why the `needs feedback` is there, as we didn't take a final decision on the approach to fixing this yet.

I hope that clarifies it.

**Comment by DB-1984** [comment]
Forgive me if I'm skewing any etiquette by posting this in this fashion, but I have a workaround (albeit an admittedly clunky one), which is working for me.

In essence, I'm utilising a separate meta query within the category loop and, in combination with some hackneyed jQuery, returning two lists of values: one containing the unique slug of each custom product attribute assigned to products within the current category, and another pulled straight from the DOM, in the form of the individual data attributes of each of the block filters (which also happen to be the attribute slug).

By comparing the two lists (in my case, on the click of the 'Filter' element which triggers the dropdown), I can then eliminate the parent element of any filter block column containing a non-matched attribute ID.

Here's my page of code - I hope it's useful, and I fully welcome any refinement!!

```
add_action('woocommerce_before_shop_loop', function(){
	
$currentCategory = get_queried_object();
$category_id = $currentCategory->term_id;

$args = array(
    'post_type'      => 'product',
    'post_status'    => 'publish',
    'posts_per_page' => -1,
    'tax_query'      => array(
        array(
            'taxonomy' => 'product_cat',
            'field'    => 'term_id',
            'terms'    => $category_id,
        ),
    ),
);

$products = new WP_Query( $args );

if ( $products->have_posts() ) {
    while ( $products->have_posts() ) {
        $products->the_post();

        $product_id = get_the_ID();

        // Retrieve attribute slugs for this product
        $attribute_slugs = array();

        $attribute_taxonomies = wc_get_attribute_taxonomies();

        if ( ! empty( $attribute_taxonomies ) ) {
            foreach ( $attribute_taxonomies as $attribute_taxonomy ) {
                $attribute = 'pa_' . $attribute_taxonomy->attribute_name;
                $terms = wp_get_post_terms( $product_id, $attribute );
                if ( ! empty( $terms ) && ! is_wp_error( $terms ) ) {
                    $slugs = wp_list_pluck( $terms, 'slug' );
                    foreach ( $slugs as $slug ) {
                        $attribute_slugs[] = $slug;
                    }
                }
            }
        }
	}

if ( ! empty( $attribute_slugs ) ) {
    $attribute_slugs_output = implode( ',', $attribute_slugs );
	//var_dump($attribute_slugs_output);
	}
}
?><script>
jQuery(document).ready(function($){
    var attribute_slugs = "<?php echo $attribute_slugs_output; ?>";
    var attribute_slugs_array = attribute_slugs.split(',');
    var attribute_slugs_values = [];

    $(attribute_slugs_array).each(function(index, value) {
        var attribute_slug_text = jQuery.trim(value);
        attribute_slugs_values.push(attribute_slug_text);
    });
	$("div.filter").click(function(){;
    var wc_block_components_values = [];

    if ($('.wc-block-components-checkbox').length) {
        $('.wc-block-components-checkbox').each(function() {
            var id = $(this).find('input').attr('id');
            wc_block_components_values.push(id);
        });

        // Compare the two arrays
        var should_hide = false;
        if (attribute_slugs_values.length !== wc_block_components_values.length) {
            should_hide = true;
        } else {
            for (var i = 0; i < attribute_slugs_values.length; i++) {
                if (!wc_block_components_values.includes(attribute_slugs_values[i])) {
                    should_hide = true;
                    break;
                }
            }
        }

        // Hide the closest parent element to the label element of the elements not present in attribute_slugs_output
        if (should_hide) {
            $('.wc-block-components-checkbox').each(function() {
                var id = $(this).find('input').attr('id');
                if (!attribute_slugs_values.includes(id)) {
                   $(this).closest("div.column").hide();
                }
            });
        }
    }
});
});
</script>
<?php });
```

**Comment by wpsoul** [comment]
Half year with no fix. I think it will be better to make filter block as server side rendering. more easy for customization, more easy for fixing and it's faster

**Comment by sunyatasattva** [comment]
@wpsoul We do agree with you, and we are working to make that happen! We don't want to use jQuery, we want a more permanent solution, so we will rely on the Interactivity API coming soon on WordPress core. If you haven't heard about it, read more here: https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/

**Comment by wpsoul** [comment]
> @wpsoul https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/

looks promising. But for now, I need to find alternative solution for product filters. Current filter block is almost unusable for production

**Comment by chyvak1831** [comment]
> > @wpsoul https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/
> 
> looks promising. But for now, I need to find alternative solution for product filters. Current filter block is almost unusable for production

I'm using classic filters for now. I created simple empty block (it's just saying 'Here displays filters' in wordpress admin on page), but on front I output classic php filter template via renderCallback.

The simplest and stable solution which I see to use correct filters (classic for now) in block theme.

**Comment by wpsoul** [comment]
@chyvak1831 yes, classic widgets are working great in classic themes, but I need for FSE. I think to make block which renders output as classic widget. If you know such ready solution, let me know

**Comment by sauls123** [comment]
Hi - is there any update on this one? Quite an important issue from a UX + Conversion Rate perspective.

**Comment by thompalstra** [comment]
Any progress yet? I've noticed a lot of customers requesting to use block themes but since nobody can use filter by attributes combined with the products (beta) block, most of these customers are starting to refrain from using woo.

**Comment by normana10** [comment]
Not _exactly_ the "Filter by Attribute" block, but I've put something together that fixes the "Filter by Price" block here #50341

The backend endpoints for the _price range summarization_ actually *already* support category filtering out of the box, so it was a simple case of getting the block to pass along the category ID and then it's good to go

That said, I'd like to get the "Filter by Attribute" block fixed up as well

That one will be a bit tricker as I _believe_ the term counts are stored in the `wp_term_taxonomy` table and are computed across _all_ products, so we'll need to compute (and cache) the results on the fly (similar to @DB-1984's solution above)

Though with the attribute filter block currently being a _static_ block, I suppose we're less concerned about any processing overhead there because at least the page will have rendered for the customer while we go fetch the relevant attributes asynchronously?

("High Performance Product Storage" when? 😜)
